### ISSUE TITLE
mwan3: Router becomes unreachable from LAN is all WANs are down at the boot time

### ISSUE THREAD

**Issue by gentoo-root** [openwrt/packages#20779]
Maintainer: @feckert
Environment: ipq806x, Netgear R7800, OpenWrt from master (https://github.com/openwrt/openwrt/commit/0dc5fc8fa53be55c3adc59055e7f6f150a5ab1ad).

Description:

1. Make sure that all WANs are offline (in fact, I performed the debugging with only one WAN, but the issue exists with multiple WANs as well).
2. Turn on the router.
3. After it boots, you can ping the router and SSH into it from LAN.
4. After a while, exactly when mwan3 moves the interface to offline (`mwan3track[1532]: Interface wan (eth0.2) is offline`), the issue appears.
5. No new connections are accepted at this point, but the established connections stay alive. For example, if you have active SSH sessions or ping in progress, it will continue, but new SSH connections will fail (the client never receives a SYN-ACK), and new instances of ping don't receive any replies.

I debugged the issue and found what's wrong at startup. At step 3, the mwan3_connected_ipv4 ipset is not populated, it contains only 224.0.0.0/3. At the same time, the mwan3_ifaces_in iptables chain is empty, so no one assigns fwmarks, and the "unreachable" rules are not triggered.

When mwan3 detects that the WAN interface is offline (step 4), it creates the mwan3_ifaces_in and mwan3_iface_in_wan chains, but it doesn't populate the mwan3_connected_ipv4 ipset, and all ingress packets from LAN belonging to new connections (only new connections hit mwan3_ifaces_in) start being dropped by the ip rule `2062:	from all fwmark 0x3e00/0x3f00 unreachable`.

During normal operation, the mwan3_connected_ipv4 ipset contains the IP address of the router (along with the LAN subnet), so new packets don't get marked to be dropped.

I believe, this should be fixed so that the mwan3_connected_ipv4 ipset is populated from the start, and the mwan3_ifaces_in* chains should also exist from the start.

The issue only happens on boot. If you manage to SSH into the router during that small time frame when it's still reachable, and restart mwan3, it will stay reachable, even though the WAN is down.

**Comment by anyuta1166** [comment]
Interesting issue. I was hardly investigating some issues with mwan3 last days and decided to check this too.

I think that mwan3_ifaces_in and mwan3_iface_in_wan chains have nothing to do with it. These chains are only for incoming traffic from wan.

The possible cause is that if mwan3_connected_ipv4 ipset is empty, then incoming connections from lan should hit the default ipv4 rule, which is mwan3_policy_balanced in your case:

```
Chain mwan3_rules (1 references)
 pkts bytes target     prot opt in     out     source               destination         
    3   252 mwan3_policy_balanced  all  --  any    any     anywhere             anywhere             mark match 0x0/0x3f00
```

Since there is no wan connection and, as I guess, last resort for this policy is set to unreachable, all connections hitting this policy are marked 0x3e00:

```
Chain mwan3_policy_balanced (2 references)
 pkts bytes target     prot opt in     out     source               destination         
    3   252 MARK       all  --  any    any     anywhere             anywhere             mark match 0x0/0x3f00 /* unreachable */ MARK xset 0x3e00/0x3f00
```

and rejected by the corresponding ip rule.

However I cannot reproduce the issue with mwan3_connected_ipv4 ipset being empty.

I've checked 2 cases: 1) all wan interfaces physically disconnected 2) all wan interfaces have cable connected, but no internet connection. Booting the router withj this conditions results in correcly populated mwan3_connected_ipv4 ipset, containing the record of lan network (192.168.0.0/24 in my case).

As a workaround you can set last resort for the mwan3_policy_balanced policy to default (use main routing table) instead of unreachable.

**Comment by anyuta1166** [comment]
Update: I was able to reproduce the issue.

I was using an older build of openwrt with mwan3 (2.10.13) and everything worked fine.

After updating to the latest openwrt 22.03-snapshot with 2.11.4 I've encountered this issue.

Indeed, when all wan are disconnected, then mwan3_connected_ipv4 is empty and connections to the router are blocked.

Also I've noticed a lot of redundant entries in mwan3_connected_ipv4 when wan IS connected.

For example:

in 2.10.13 it has one entry for load network:
```
192.168.0.0/24
```

but in 2.11.4 it has 4 entries:
```
192.168.0.255
192.168.0.0
192.168.0.0/24
192.168.0.254
```

It seems that something was broken between 2.10.13 and 2.11.4.
